### PR TITLE
feat: Add support for Hubble control plane in Retina agent

### PR DESCRIPTION
**Pull request by anubhabMajumdar**
## Description

This PR adds support for Hubble control plane in Retina agent. This is being done in the most backward compatible way possible. 

I am adding a new subcommand called `hubble-control-plane` which will start Hubble instead of existing control plane.

```
$  ./retina -h                                                                                                                                 ✔  took 5s 
Start Retina Agent

Usage:
  retina-agent [flags]
  retina-agent [command]

Available Commands:
  completion           Generate the autocompletion script for the specified shell
  help                 Help about any command
  hubble-control-plane Start Hubble control plane

Flags:
      --config string                      config file (default "/retina/config/config.yaml")
      --health-probe-bind-address string   The address the probe endpoint binds to. (default ":18081")
  -h, --help                               help for retina-agent
      --leader-elect                       Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.
      --metrics-bind-address string        The address the metric endpoint binds to. (default ":18080")

Use "retina-agent [command] --help" for more information about a command.
$
$
$  ./retina                                                                                                                                                ✔
Starting Retina Agent
starting Retina daemon with legacy control plane
...
$
$
$ ./retina hubble-control-plane -h
Start Hubble control plane

Usage:
  retina-agent hubble-control-plane [flags]
  retina-agent hubble-control-plane [command]

Available Commands:
  hive        Inspect the hive

Flags:
      --cluster-name string                        name of the cluster (default "default")
...
$
$
$ ./retina hubble-control-plane --config-dir .
ts=2024-06-13T18:20:47.888Z level=info caller=hubble/daemon_main.go:276 msg="Traces telemetry initialized with zapai" version= appInsightsID=
time="2024-06-13T18:20:47Z" level=info msg=Invoked duration="483.869µs" function="pprof.init.func1 (pkg/pprof/cell.go:49)" subsys=hive
time="2024-06-13T18:20:47Z" level=info msg=Invoked duration="77.011µs" function="gops.registerGopsHooks (pkg/gops/cell.go:38)" subsys=hive
time="2024-06-13T18:20:47Z" level=info msg=Invoked duration=10.503879ms function="github.com/microsoft/retina/cmd/hubble.init.func3 (cmd/hubble/daemon.go:68)" subsys=hive
time="2024-06-13T18:20:47Z" level=info msg="&{{ 0}  [] 0s false true false false false false}" subsys=agent-config
time="2024-06-13T18:20:47Z" level=info msg="configuring telemetry" app-insights-id= retina-version= subsys=telemetry
time="2024-06-13T18:20:47Z" level=info msg="telemetry disabled" subsys=telemetry
ts=2024-06-13T18:20:47.902Z level=info caller=metrics/metrics.go:169 msg="Metrics initialized"
...
```

## Changes made

- Adopting [Hive](https://docs.cilium.io/en/latest/contributing/development/hive/) for dependency injection
- Moving to [Cobra](https://github.com/spf13/cobra) for CLI
- `controller/main.go` now is just the starting point of the command
- `retina/cmd` now houses `rootCmd` (starts retina as is) and `hubble` (starts Hubble control plane)
- In terms of Dockerfiles, packaging `Hubble` cli in agent image
- Adding new YAML files to install Retina with Hubble
- Moved the current YAML files under `deploy/legacy`  
- Fix the links in `doc`
- Update cilium version to pull in commits from upstream needed for starting Hubble
- Update `init` to add a step that creates Cilium dirs (This will happen for current control plane as well, but it consumes no resources, just creates an empty directory)
- All new packages under `pkg` contains business logic required to run Hubble (node reconciler, Hubble control plane, IPCache, etc.)
- Minor changes to `test/e2e` to support change to deployment directory (`deploy` -> `deploy/legacy`)

## Related Issue

https://github.com/microsoft/retina/issues/418

## Checklist

- [x] I have read the [contributing documentation](https://retina.sh/docs/contributing).
- [x] I signed and signed-off the commits (`git commit -S -s ...`). See [this documentation](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) on signing commits.
- [x] I have correctly attributed the author(s) of the code.
- [x] I have tested the changes locally.
- [x] I have followed the project's style guidelines.
- [x] I have updated the documentation, if necessary.
- [x] I have added tests, if applicable.

### Retina with Hubble

image

### Retina

image
---

Please refer to the [CONTRIBUTING.md](../CONTRIBUTING.md) file for more information on how to contribute to this project.

### CODE REVIEW

File: cmd/legacy/subcmd.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+package legacy
```
Comment (matmerr) [inline cmd/legacy/subcmd.go:3]: IMO legacy is a bit abstract, and when legacy is the only option it can be confusing. Can we refer to this as a specific control plane type? i.e. teams vs teams(new)

File: Makefile [hunk 2]
```diff
@@ -463,6 +520,11 @@ quick-build:
...
 
+.PHONY: quick-deploy-hubble
+quick-deploy-hubble:
+	$(MAKE) helm-uninstall || true
+	$(MAKE) helm-install-without-tls HELM_IMAGE_TAG=$(TAG)-linux-amd64
+
 .PHONY: simplify-dashboards
 simplify-dashboards:
-	cd deploy/grafana/dashboards/ && go test . -tags=dashboard,simplifydashboard -v
+	cd deploy/legacy/grafana/dashboards/ && go test . -tags=dashboard,simplifydashboard -v
```
Comment (huntergregory) [inline Makefile:530]: can we run this for all dashboards?

File: deploy/legacy/graphana/dashboards/README.md
Comment (huntergregory) [inline deploy/legacy/graphana/dashboards/README.md:1]: This description within this file is off now (these are outdated with Grafana, and the new dashboards in this PR are actually up to date with Grafana)

File: deploy/legacy/graphana/dashboards/clusters.json
Comment (huntergregory) [inline deploy/legacy/graphana/dashboards/clusters.json:1]: Some dashboard suggestions:
- Is "legacy" the right term for these? Legacy for manifests means outdated, but new manifests still support non-hubble. These dashboards are still for non-hubble. Maybe just rename to like dns-non-hubble.json and pod-level-non-hubble.json?
- No need to have two files for clusters.json (the new one is just an updated version of the old)

File: deploy/legacy/graphana/dashboards/simplify-grafana.go
Comment (huntergregory) [inline deploy/legacy/graphana/dashboards/simplify-grafana.go:1]: these "simplify" files ought to be used still. Keep outside of legacy folder?

File: pkg/shared/config/type.go
Comment (huntergregory) [inline pkg/shared/config/type.go:1]: nit: delete empty file?

File: pkg/shared/telemetry/type.go
Comment (huntergregory) [inline pkg/shared/telemetry/type.go:1]: nit: delete empty file?

File: pkg/monitoragent/cell_linux.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+var (
+	Cell = cell.Module(
+		"monitor-agent",
+		"Consumes the cilium events map and distributes those and other agent events",
+
+		cell.Provide(newMonitorAgent),
+		cell.Config(defaultConfig),
+	)
+
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "monitor-agent")
```
Comment (huntergregory) [inline pkg/monitoragent/cell_linux.go:28]: nit: is this used? Seeing a parameter for `Log` below too

File: cmd/hubble/subcmd_linux.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	"fmt"
+	"os"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/spf13/cobra"
+	"go.etcd.io/etcd/version"
+)
+
+func Cmd(agentHive *hive.Hive) *cobra.Command {
```
Comment (huntergregory) [inline cmd/hubble/subcmd_linux.go:15]: nit: is this used anywhere?

**Review by huntergregory — COMMENTED** [review]
Looks awesome. I mainly had nitpicks about dashboards and deleting unused bits

File: pkg/monitoragent/monitoragent_linux.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+		newListener.Close()
+		return
+	}
+
+	version := newListener.Version()
+	switch newListener.Version() { //nolint:exhaustive // the only other case is unsupported which is covered by default
+	case listener.Version1_2:
+		a.listeners[newListener] = struct{}{}
+	default:
+		newListener.Close()
```
Comment (matmerr) [inline pkg/monitoragent/monitoragent_linux.go:120]: same here, exposing close behavior in an exposed register function could result in unexpected behavior by function name

File: .golangci.yaml [hunk 6]
```diff
@@ -12,6 +12,7 @@ linters:
   disable:
   - maligned
   - scopelint
+  - gci
```
Comment (matmerr) [inline .golangci.yaml:15]: why this change? imo definitely helps with formatting

File: cmd/hubble/subcmd_linux.go (hunk 4 above)
Comment (anubhabMajumdar) [reply cmd/hubble/subcmd_linux.go:15]: Removed.

File: pkg/monitoragent/cell_linux.go (hunk 3 above)
Comment (anubhabMajumdar) [reply pkg/monitoragent/cell_linux.go:28]: Yes, there are references to this variable in this file.

File: pkg/shared/telemetry/type.go
Comment (anubhabMajumdar) [reply pkg/shared/telemetry/type.go:1]: Added a comment as to why these are required - Windows build and lint.

File: Makefile (hunk 2 above)
Comment (anubhabMajumdar) [reply Makefile:530]: Good idea.

File: deploy/legacy/graphana/dashboards/clusters.json
Comment (anubhabMajumdar) [reply deploy/legacy/graphana/dashboards/clusters.json:1]: We are sticking to legacy and Hubble as two control plane options. Also, I want to keep the two separated so that if we deprecate one, we just delete the particular directories.

File: deploy/legacy/graphana/dashboards/simplify-grafana.go
Comment (anubhabMajumdar) [reply deploy/legacy/graphana/dashboards/simplify-grafana.go:1]: Maintaining two sets.

File: pkg/plugin/dropreason/kprobe_bpfel_x86.o
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply pkg/plugin/dropreason/kprobe_bpfel_x86.o:1]: I will clean them up in a separate change.

File: cmd/hubble/subcmd_linux.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	hubbleCmd := &cobra.Command{
+		Use:   "hubble-control-plane",
```
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply cmd/hubble/subcmd_linux.go:17]: It cannot be a flag, it's very different than current control plane and requires all new flags and setup. IMO, a new sub command makes sense. And because it's. sub-command, I kept it a noun.

File: deploy/hubble/manifests/controller/helm/retina/templates/operator/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                    values:
+                      - linux
+      securityContext:
+        runAsNonRoot: true
+        runAsUser: 1000
+      containers:
+        - command:
+            - /retina-operator
+          args:
+            - --config-dir=/retina
```
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply deploy/hubble/manifests/controller/helm/retina/templates/operator/deployment.yaml:53]: `Hubble` expects the directory, so we need to provide the directory path and not the file.

File: pkg/k8s/watcher_linux.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+	})
+	return w, nil
+}
+
+func Start(ctx context.Context, k *watchers.K8sWatcher) {
+	logger.Info("Starting Kubernetes watcher")
+
+	option.Config.K8sSyncTimeout = 3 * time.Minute //nolint:gomnd // this duration is self-explanatory
+	syncdCache := make(chan struct{})
+	go k.InitK8sSubsystem(ctx, k8sResources, []string{}, syncdCache)
```
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply pkg/k8s/watcher_linux.go:88]: This is how the `InitK8sSubsystem` is designed. Since the function doesn't return error, the closing of `syncdCache` is the only indication that everything went OK and that watchers are running for all the resources we want.

File: pkg/monitoragent/monitoragent_linux.go [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+	if a == nil || newListener == nil {
+		return
+	}
+
+	a.Lock()
+	defer a.Unlock()
+
+	if isCtxDone(a.ctx) {
+		log.Debug("RegisterNewListener called on stopped monitor")
+		newListener.Close()
```
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply pkg/monitoragent/monitoragent_linux.go:111]: This is the API contract between Hubble and the "Monitor Agent" it expects.

File: pkg/monitoragent/monitoragent_linux.go (hunk 5 above)
Comment (anubhabMajumdar) [reply pkg/monitoragent/monitoragent_linux.go:120]: Same as https://github.com/microsoft/retina/pull/432#discussion_r1649442249

File: .golangci.yaml (hunk 6 above)
Comment (anubhabMajumdar) [reply .golangci.yaml:15]: @timraymond and I were discussing removing this as it messes up imports ordering in a way that's not necessarily "better" in our opinion. But open to discussion.

File: pkg/monitoragent/monitoragent_linux.go [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+	ml.Close()
+}
+
+func (a *monitorAgent) RegisterNewConsumer(newConsumer consumer.MonitorConsumer) {
+	if a == nil || newConsumer == nil {
+		return
+	}
+
+	if isCtxDone(a.ctx) {
+		log.Debug("RegisterNewConsumer called on stopped monitor")
```
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply pkg/monitoragent/monitoragent_linux.go:153]: The daemon creates the agent and Hubble. If the daemon cancels the context, the agent doesn't need to accept new consumers.

File: pkg/hubble/hubble_linux.go [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+		ipc:            params.IPCache,
+		nodeReconciler: params.NodeReconciler,
+	}
+	rh.log.Logger.SetLevel(logrus.InfoLevel)
+
+	return rh
+}
+
+func (rh *RetinaHubble) defaultOptions() {
+	// Set the default values for Retina-enterprise.
```
Comment (matmerr) [inline pkg/hubble/hubble_linux.go:65]: 👀

File: pkg/config/hubble_config.go [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+	flags.String("metrics-bind-address", c.MetricsBindAddress, "The address the metric endpoint binds to.")
+	flags.String("health-probe-bind-address", c.HealthProbeBindAddress, "The address the probe endpoint binds to.")
+
+	flags.Bool("leader-election", c.LeaderElection, "Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.")
+	flags.String("cluster-name", c.ClusterName, "name of the cluster")
+}
+
+var (
+	DefaultRetinaHubbleConfig = RetinaHubbleConfig{
+		MetricsBindAddress:     ":18000",
```
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply pkg/config/hubble_config.go]: The 10093 port is set through `apiserver.port` (hard disagree with the name). This is a placeholder for a required flag.

**Review by matmerr — DISMISSED** [review]
🛰️

File: Makefile [hunk 14]
```diff
@@ -527,4 +527,4 @@ quick-deploy-hubble:
 
 .PHONY: simplify-dashboards
 simplify-dashboards:
-	cd deploy/legacy/grafana/dashboards/ && go test . -tags=dashboard,simplifydashboard -v
+	go test -tags=dashboard,simplifydashboard -v ./deploy/...
```
Comment (huntergregory) [inline Makefile:530]: The test logic actually modifies JSON files in the current directory. Could we cd into both directories?

**Review by huntergregory — CHANGES_REQUESTED** [review]
It looks like the UT files are missing in deploy/hubble/grafana/ (can copy from deploy/legacy/grafana/)

**Review by huntergregory — DISMISSED** [review]

**Review by huntergregory — APPROVED** [review]

**Review by huntergregory — DISMISSED** [review]

**Review by rbtr — DISMISSED** [review]

**Review by rbtr — DISMISSED** [review]

**Review by rbtr — APPROVED** [review]
